funsor/joint.py

```diff
@@ -0,0 +1,239 @@
+            assert (reduced_vars - gaussian_vars).issubset(d.name for d in self.deltas)
+
+            # Integrate out delayed degenerate variables, i.e. drop them.
+            deltas = tuple(d for d in self.deltas if d.name not in reduced_vars)
+
+            assert not lazy_vars
+            return (Joint(deltas, discrete) + gaussian).reduce(ops.logaddexp, lazy_vars)
+
+        if op is ops.add:
+            raise NotImplementedError('TODO product-reduce along a plate dimension')
+
+        return None  # defer to default implementation
+
+
+@eager.register(Joint, tuple, Funsor, Funsor)
+def eager_joint(deltas, discrete, gaussian):
+    # Demote a Joint to a simpler elementart funsor.
+    if not deltas:
+        if gaussian is Number(0):
+            return discrete
+        elif discrete is Number(0):
+            return gaussian
+    elif len(deltas) == 1:
+        if discrete is Number(0) and gaussian is Number(0):
+            return deltas[0]
+
+    return None  # defer to default implementation
+
+
+################################################################################
+# Patterns to update a Joint with other funsors
+################################################################################
+
+@eager.register(Binary, AddOp, Joint, Joint)
+def eager_add(op, joint, other):
+    # Fuse two joint distributions.
+    for d in other.deltas:
+        joint += d
+    joint += other.discrete
+    joint += other.gaussian
+    return joint
+
+
+@eager.register(Binary, AddOp, Joint, Delta)
+def eager_add(op, joint, delta):
+    # Update with a degenerate distribution, typically a monte carlo sample.
+    if delta.name in joint.inputs:
+        joint = joint.eager_subs(((delta.name, delta.point),))
+        if not isinstance(joint, Joint):
+            return joint + delta
+    for d in joint.deltas:
+        if d.name in delta.inputs:
+            delta = delta.eager_subs(((d.name, d.point),))
+    deltas = joint.deltas + (delta,)
+    return Joint(deltas, joint.discrete, joint.gaussian)
+
+
+@eager.register(Binary, AddOp, Joint, (Number, Tensor))
+def eager_add(op, joint, other):
+    # Update with a delayed discrete random variable.
+    subs = tuple((d.name, d.point) for d in joint.deltas if d in other.inputs)
+    if subs:
+        return joint + other.eager_subs(subs)
+    return Joint(joint.deltas, joint.discrete + other, joint.gaussian)
+
+
+@eager.register(Binary, AddOp, Joint, Gaussian)
+def eager_add(op, joint, other):
+    # Update with a delayed gaussian random variable.
+    subs = tuple((d.name, d.point) for d in joint.deltas if d in other.inputs)
+    if subs:
+        other = other.eager_subs(subs)
+    if joint.gaussian is not Number(0):
+        other = joint.gaussian + other
+    if not isinstance(other, Gaussian):
+        return Joint(joint.deltas, joint.discrete) + other
+    return Joint(joint.deltas, joint.discrete, other)
+
+
+@eager.register(Binary, AddOp, Joint, Binary)
```

Comment:
How would this case come up in `eager`?

Reply:
Oh good point, it wouldn't come up. I've removed this case.